src/main/scala/org/broadinstitute/hail/variant/VariantSampleMatrix.scala

```diff
@@ -212,24 +212,26 @@ case class VSMSubgen[T](
   saSigGen: Gen[Type],
   vaSigGen: Gen[Type],
   globalSigGen: Gen[Type],
-  saGen: (Int, Type) => Gen[IndexedSeq[Annotation]],
+  saGen: (Type) => Gen[Annotation],
   vaGen: (Type) => Gen[Annotation],
   globalGen: (Type) => Gen[Annotation],
   vGen: Gen[Variant],
   tGen: (Variant) => Gen[T]) {
 
   def gen(sc: SparkContext)(implicit tct: ClassTag[T]): Gen[VariantSampleMatrix[T]] =
-    for (vaSig <- vaSigGen;
-         saSig <- saSigGen;
-         globalSig <- globalSigGen;
-         sampleIds <- sampleIdGen;
-         global <- globalGen(globalSig);
-         saValues <- saGen(sampleIds.length, saSig);
+    for (subsizes <- Gen.getPartition(7);
+         vaSig <- vaSigGen.resize(subsizes(0));
+         saSig <- saSigGen.resize(subsizes(1));
+         globalSig <- globalSigGen.resize(subsizes(2));
+         sampleIds <- sampleIdGen.resize(subsizes(3));
+         global <- globalGen(globalSig).resize(subsizes(4));
+         saValues <- Gen.buildableOfN(sampleIds.length, saGen(saSig)).resize(subsizes(5));
          rows <- Gen.distinctBuildableOf[Seq[(Variant, (Annotation, Iterable[T]))], (Variant, (Annotation, Iterable[T]))](
-           for (v <- vGen;
-                va <- vaGen(vaSig);
-                ts <- Gen.buildableOfN[Iterable[T], T](sampleIds.length, tGen(v)))
-             yield (v, (va, ts))))
+           for (subsubsizes <- Gen.getPartition(3);
+                v <- vGen.resize(subsubsizes(0));
+                va <- vaGen(vaSig).resize(subsubsizes(1));
+                ts <- Gen.buildableOfN[Iterable[T], T](sampleIds.length, tGen(v)).resize(subsubsizes(2)))
+             yield (v, (va, ts))).resize(subsizes(6)))
```

Comment:
This is very nice, but I still think it is O(n^2).  sampleIds is O(n) (subsize 3), rows is O(n) (subsize 6), and since you're calling into buildableOfN in the row generator, you're effectively circumventing the size constraint, each row is O(n) instead of total size constrained by subsize 6.

I don't see how to do it without a division.  I'd divide subsize 6 by the number of samples and I think you're golden.  Also, you haven't addressed the quadratic terms coming from the annotations.